File: app/src/main/java/com/kamelong/OuDia/TimeTableSorter.java
package com.kamelong.OuDia;


import com.kamelong.tool.SDlog;

import java.util.ArrayList;
import java.util.Collections;

/*
 * Copyright (c) 2019 KameLong
 * contact:kamelong.com
 *
 * This source code is released under GNU GPL ver3.
 */

/**
 * 時刻表並び替え作業を行う作業クラスです。
 * 時刻表並び替えが終了したらこのオブジェクトを破棄してください。
 */
public class TimeTableSorter {
    LineFile lineFile;
    Train[] trainList;
    int direction;
    private ArrayList<Integer> sortBefore;
    private ArrayList<Integer> sortAfter;
    boolean[] sorted;
    int loopNum=0;

    /**
     * ソートする時刻表を入力します。
     */
    public TimeTableSorter(LineFile lineFile,Train[] trains,int direction){
        this.lineFile=lineFile;
        this.trainList =trains;
        this.direction=direction;
        this.sortBefore =new ArrayList<>();
        this.sortAfter =new ArrayList<>();
        for(int i=0;i<trains.length;i++){
            sortBefore.add(i);
        }
        sorted=new boolean[lineFile.getStationNum()];
        for(int i=0;i<sorted.length;i++){
            sorted[i]=false;
        }

    }

    /**
     * 列車番号を(数字)+(アルファベットなど)と分離する
     * まずはアルファベットでソートして、その後数字でソートする
     */
    public ArrayList<Train>sortNumber(){
        ArrayList<TrainNumberSorter>sorter=new ArrayList<>();
        for(Train train:trainList){
            sorter.add(new TrainNumberSorter(train));
        }
        Collections.sort(sorter);
        ArrayList<Train>result=new ArrayList<>();
        for(TrainNumberSorter sort:sorter){
            result.add(sort.train);
        }
        return result;
    }
    /**
     * 列車番号を(数字)+(アルファベットなど)と分離する
     * まずはアルファベットでソートして、その後数字でソートする
     */
    public ArrayList<Train>sortType(){
        ArrayList<TrainTypeSorter>sorter=new ArrayList<>();
        for(Train train:trainList){
            sorter.add(new TrainTypeSorter(train));
        }
        Collections.sort(sorter);
        ArrayList<Train>result=new ArrayList<>();
        for(TrainTypeSorter sort:sorter){
            result.add(sort.train);
        }
        return result;
    }
    /**
     * 列車番号を(数字)+(アルファベットなど)と分離する
     * まずはアルファベットでソートして、その後数字でソートする
     */
    public ArrayList<Train>sortName(){
        ArrayList<TrainNameSorter>sorter=new ArrayList<>();
        for(Train train:trainList){
            sorter.add(new TrainNameSorter(train));
        }
        Collections.sort(sorter);
        ArrayList<Train>result=new ArrayList<>();
        for(TrainNameSorter sort:sorter){
            result.add(sort.train);
        }
        return result;
    }
    /**
     * 列車番号を(数字)+(アルファベットなど)と分離する
     * まずはアルファベットでソートして、その後数字でソートする
     */
    public ArrayList<Train>sortRemark(){
        ArrayList<TrainRemarkSorter>sorter=new ArrayList<>();
        for(Train train:trainList){
            sorter.add(new TrainRemarkSorter(train));
        }
        Collections.sort(sorter);
        ArrayList<Train>result=new ArrayList<>();
        for(TrainRemarkSorter sort:sorter){
            result.add(sort.train);
        }
        return result;
    }

    /**
     * 列車時刻でソートします。
     * @param stationIndex ソート基準時刻
     */
    public ArrayList<Train> sort(int stationIndex){
        ArrayList<Train> result = new ArrayList<>();

        try {
            loopNum = 0;
            for (int i = 0; i < sortBefore.size(); i++) {
                if (trainList[sortBefore.get(i)].getPredictionTime(stationIndex) > 0 && !trainList[sortBefore.get(i)].checkDoubleDay()) {
                    //今からsortAfterに追加する列車の基準駅の時間
                    int baseTime = trainList[sortBefore.get(i)].getPredictionTime(stationIndex);
                    int j;
                    for (j = sortAfter.size(); j > 0; j--) {
                        if (trainList[sortAfter.get(j - 1)].getPredictionTime(stationIndex) < baseTime) {
                            break;
                        }
                    }
                    sortAfter.add(j, sortBefore.get(i));
                    sortBefore.remove(i);
                    i--;
                }
            }
            sorted[stationIndex] = true;
            if (direction == Train.DOWN) {
                sortDown(stationIndex);
            } else {
                sortUp(stationIndex);
            }
            sortAfter.addAll(sortBefore);
            for (int i : sortAfter) {
                result.add(trainList[i]);
            }
            return result;
        }catch (Exception e){
            SDlog.log(e);
        }
        result=new ArrayList<>();
        for (int i : sortBefore) {
            result.add(trainList[i]);
        }
        return result;
    }

    /**
     * 路線内を上り方向に探索していきます。
     */
    public void sortUp(int stationIndex)throws Exception{
        loopNum++;
        if(loopNum>50){
            SDlog.toast("エラーこのダイヤファイルの路線分岐が複雑であるため、列車の並び替え時に無限ループに陥りました。並び替え操作を強制終了します");
            throw new Exception("並び替えエラー："+lineFile.name);
        }

            boolean skip = true;//ソート済みの路線から外れ、別の路線に入る場合skipfragがtrueになる。//ソート済み領域に戻ればskip=false

            for (; stationIndex >=0; stationIndex--) {
                //上り方向に探索
                if (sorted[stationIndex]) {
                    skip = false;
                }
                if (lineFile.getStation(stationIndex).brunchCoreStationIndex >= 0) {
                    //この駅が分岐駅設定されている場合
                    if (lineFile.getStation(stationIndex).brunchCoreStationIndex > stationIndex) {
                        //下から分岐する場合はソート対象外
                        skip = true;
                    }
                }

                if (!sorted[stationIndex]) {
                    //この駅がまだソートされていないとき
                    if(skip) {
                        boolean subskip=true;
                        //この駅がスキップされるとき
                        if(lineFile.getStation(stationIndex).brunchCoreStationIndex>=0) {
                            //分岐駅設定あり
                            if(sorted[lineFile.getStation(stationIndex).brunchCoreStationIndex]) {
                                //分岐元がソート済み
                                //ソートする
                                subskip=false;
                            }
                        }
                        //この駅がどこかの分岐元でその駅がソートされている可能性
                        for(int i=0;i<lineFile.getStationNum();i++) {
                            if(lineFile.getStation(i).brunchCoreStationIndex==stationIndex&&sorted[i]) {
                                subskip=false;
                            }
                        }
                        if(subskip) {
                            //スキップする
                            continue;
                        }
                    }
                    //ソートする
                    System.out.println("sortUp:\t" + stationIndex + "\t(" + lineFile.getStation(stationIndex).name + ")");
                    ArrayList<Integer> stations = new ArrayList<>();
                    stations.add(stationIndex);
                    if (lineFile.getStation(stationIndex).brunchCoreStationIndex >= 0) {
                        stations.add(lineFile.getStation(stationIndex).brunchCoreStationIndex);
                    }
                    for (int i = 0; i < lineFile.getStationNum(); i++) {
                        if (lineFile.getStation(i).brunchCoreStationIndex == stationIndex) {
                            stations.add(i);
                        }
                    }
                    if (direction == Train.DOWN) {
                        addTrainInSort1(stations);
                    } else {
                        addTrainInSort2(stations);
                    }
                    sorted[stationIndex] = true;
                    skip=false;
                }

                if (lineFile.getStation(stationIndex).brunchCoreStationIndex >= 0) {
                    //この駅が分岐駅設定されている場合
                    if (lineFile.getStation(stationIndex).brunchCoreStationIndex < stationIndex) {
                        //上へ分岐するときは次の駅からソート対象外
                        skip = true;
                    }
                }
            }

            //ループが終わったのに、既にソート済みの駅が最後に残った場合や最後スキップされていた場合
            for (int i = 0; i < lineFile.getStationNum(); i++) {
                if (!sorted[i]) {
                    sortDown(0);

                    return;
                }
            }
    }
    /**
     * 路線内を下り方向に探索していきます。
     */
    public void sortDown(int stationIndex)throws Exception{
        loopNum++;
        if(loopNum>50){
            SDlog.toast("エラーこのダイヤファイルの路線分岐が複雑であるため、列車の並び替え時に無限ループに陥りました。並び替え操作を強制終了します");
            throw new Exception("並び替えエラー："+lineFile.name);
        }

            boolean skip = true;//ソート済みの路線から外れ、別の路線に入る場合skipfragがtrueになる。//ソート済み領域に戻ればskip=false

            for (; stationIndex < lineFile.getStationNum(); stationIndex++) {
                //下り方向に探索
                if (sorted[stationIndex]) {
                    skip = false;
                }
                if (lineFile.getStation(stationIndex).brunchCoreStationIndex >= 0) {
                    //この駅が分岐駅設定されている場合
                    if (lineFile.getStation(stationIndex).brunchCoreStationIndex < stationIndex) {
                        //上から分岐する場合はソート対象外
                        skip = true;
                    }
                }

                if (!sorted[stationIndex]) {
                    //この駅がまだソートされていないとき
                    if(skip) {
                        boolean subskip=true;
                        //この駅がスキップされるとき
                        if(lineFile.getStation(stationIndex).brunchCoreStationIndex>=0) {
                            //分岐駅設定あり
                            if(sorted[lineFile.getStation(stationIndex).brunchCoreStationIndex]) {
                                //分岐元がソート済み
                                //ソートする
                                subskip=false;
                            }
                        }
                        //この駅がどこかの分岐元でその駅がソートされている可能性
                        for(int i=0;i<lineFile.getStationNum();i++) {
                            if(lineFile.getStation(i).brunchCoreStationIndex==stationIndex&&sorted[i]) {
                                subskip=false;
                            }
                        }
                        if(subskip) {
                            //スキップする
                            continue;
                        }
                    }
                    //ソートする
                    System.out.println("sortDown:\t" + stationIndex + "\t(" + lineFile.getStation(stationIndex).name + ")");
                    ArrayList<Integer> stations = new ArrayList<>();
                    stations.add(stationIndex);
                    if (lineFile.getStation(stationIndex).brunchCoreStationIndex >= 0) {
                        stations.add(lineFile.getStation(stationIndex).brunchCoreStationIndex);
                    }
                    for (int i = 0; i < lineFile.getStationNum(); i++) {
                        if (lineFile.getStation(i).brunchCoreStationIndex == stationIndex) {
                            stations.add(i);
                        }
                    }
                    if (direction == Train.DOWN) {
                        addTrainInSort2(stations);
                    } else {
                        addTrainInSort1(stations);
                    }
                    sorted[stationIndex] = true;
                    skip=false;
                }

                if (lineFile.getStation(stationIndex).brunchCoreStationIndex >= 0) {
                    //この駅が分岐駅設定されている場合
                    if (lineFile.getStation(stationIndex).brunchCoreStationIndex > stationIndex) {
                        //下へ分岐するときは次の駅からソート対象外
                        skip = true;
                    }
                }
            }
            //ループが終わったのに、既にソート済みの駅が最後に残った場合や最後スキップされていた場合
            for (int i = 0; i < lineFile.getStationNum(); i++) {
                if (!sorted[i]) {
                    sortUp(lineFile.getStationNum() - 1);
                    break;
                }
            }
    }
    /**
     * 列車をsortAfterに時刻前方から挿入する
     * station[0]に停車する列車がソート対象
     * station[1以上]は同一駅
     */
    private void addTrainInSort1(ArrayList<Integer> station) {
        for (int i = sortBefore.size(); i > 0; i--) {
            int baseTime = trainList[sortBefore.get(i - 1)].getTime(station.get(0),Train.ARRIVE,true);
            if (baseTime < 0 || trainList[sortBefore.get(i - 1)].checkDoubleDay()) {
                continue;
            }
            int j = 0;
            boolean frag = false;

            for (j = 0; j < sortAfter.size(); j++) {

                int sortTime=-1;
                for(int s:station){
                    sortTime=Math.max(sortTime,trainList[sortAfter.get(j)].getPredictionTime(s,Train.DEPART));
                }
                if (sortTime < 0) {
                    continue;
                }
                frag = true;
                if (sortTime >= baseTime) {
                    break;
                }
            }
            if (frag) {
                sortAfter.add(j, sortBefore.get(i - 1));
                sortBefore.remove(i - 1);
            }
        }
    }

    private void addTrainInSort2(ArrayList<Integer> station) {
        for (int i = 0; i < sortBefore.size(); i++) {
            int baseTime = trainList[sortBefore.get(i)].getDepTime(station.get(0));
            if (baseTime < 0 || trainList[sortBefore.get(i)].checkDoubleDay()) {
                continue;
            }
            int j = 0;
            boolean frag = false;

            for (j = sortAfter.size()-1; j >= 0; j--) {

                int sortTime=-1;
                for(int s:station){
                    sortTime=Math.max(sortTime,trainList[sortAfter.get(j)].getPredictionTime(s,Train.ARRIVE));
                }
                if (sortTime < 0) {
                    continue;
                }
                frag = true;
                if (sortTime <= baseTime) {
                    break;
                }
            }
            if (frag) {
                sortAfter.add(j+1, sortBefore.get(i));
                sortBefore.remove(i);
                i--;
            }

        }

    }
    static class TrainNumberSorter implements Comparable<TrainNumberSorter>{
        Train train;
        String name="";
        int number=0;
        public TrainNumberSorter(Train train){
            this.train=train;
            String value=train.number;
            for(int i=0;i<value.length();i++){
                if(!isNumber(value.toCharArray()[i])){
                    if(i!=0) {
                        number = Integer.parseInt(value.substring(0, i));
                    }
                    name=value.substring(i);
                    return;
                }
            }
            if(value.length()!=0){
                number=Integer.parseInt(value);

            }
        }

        boolean isNumber(char c){
            return c >= '0' && c <= '9';
        }

        @Override
        public int compareTo(TrainNumberSorter o) {
            if(this.name.equals(o.name)){
                return this.number-o.number;
            }
            return this.name.compareTo(o.name);

        }
    }
    static class TrainTypeSorter implements Comparable<TrainTypeSorter>{
        Train train;
        public TrainTypeSorter(Train train){
            this.train=train;
        }
        @Override
        public int compareTo(TrainTypeSorter o) {
            return this.train.type-o.train.type;
        }
    }
    static class TrainNameSorter implements Comparable<TrainNameSorter>{
        Train train;
        public TrainNameSorter(Train train){
            this.train=train;
        }
        @Override
        public int compareTo(TrainNameSorter o) {
            if(this.train.name.equals(o.train.name)){
                return this.train.count.compareTo(o.train.count);
            }
            return this.train.name.compareTo(o.train.name);
        }
    }
    static class TrainRemarkSorter implements Comparable<TrainRemarkSorter>{
        Train train;
        public TrainRemarkSorter(Train train){
            this.train=train;
        }
        @Override
        public int compareTo(TrainRemarkSorter o) {
            return this.train.remark.compareTo(o.train.remark);
        }
    }

}

File: app/src/main/java/com/kamelong/OuDia/TimeTableSorter.kt
package com.kamelong.OuDia

import com.kamelong2.aodia.SDlog
import java.util.*

/*
 * Copyright (c) 2019 KameLong
 * contact:kamelong.com
 *
 * This source code is released under GNU GPL ver3.
 */
/**
 * 時刻表並び替え作業を行う作業クラスです。
 * 時刻表並び替えが終了したらこのオブジェクトを破棄してください。
 */
class TimeTableSorter(var lineFile: LineFile, var trainList: Array<Train>, var direction: Int) {
    private val sortBefore: ArrayList<Int>
    private val sortAfter: ArrayList<Int>
    var sorted: BooleanArray
    var loopNum = 0

    /**
     * 列車番号を(数字)+(アルファベットなど)と分離する
     * まずはアルファベットでソートして、その後数字でソートする
     */
    fun sortNumber(): ArrayList<Train> {
        val sorter = ArrayList<TrainNumberSorter>()
        for (train in trainList) {
            sorter.add(TrainNumberSorter(train))
        }
        Collections.sort(sorter)
        val result = ArrayList<Train>()
        for (sort in sorter) {
            result.add(sort.train)
        }
        return result
    }

    /**
     * 列車番号を(数字)+(アルファベットなど)と分離する
     * まずはアルファベットでソートして、その後数字でソートする
     * @return
     */
    fun sortType(): ArrayList<Train> {
        val sorter = ArrayList<TrainTypeSorter>()
        for (train in trainList) {
            sorter.add(TrainTypeSorter(train))
        }
        Collections.sort(sorter)
        val result = ArrayList<Train>()
        for (sort in sorter) {
            result.add(sort.train)
        }
        return result
    }

    /**
     * 列車番号を(数字)+(アルファベットなど)と分離する
     * まずはアルファベットでソートして、その後数字でソートする
     * @return
     */
    fun sortName(): ArrayList<Train> {
        val sorter = ArrayList<TrainNameSorter>()
        for (train in trainList) {
            sorter.add(TrainNameSorter(train))
        }
        Collections.sort(sorter)
        val result = ArrayList<Train>()
        for (sort in sorter) {
            result.add(sort.train)
        }
        return result
    }

    /**
     * 列車番号を(数字)+(アルファベットなど)と分離する
     * まずはアルファベットでソートして、その後数字でソートする
     * @return
     */
    fun sortRemark(): ArrayList<Train> {
        val sorter = ArrayList<TrainRemarkSorter>()
        for (train in trainList) {
            sorter.add(TrainRemarkSorter(train))
        }
        Collections.sort(sorter)
        val result = ArrayList<Train>()
        for (sort in sorter) {
            result.add(sort.train)
        }
        return result
    }

    /**
     * 列車時刻でソートします。
     * @param stationIndex ソート基準時刻
     */
    fun sort(stationIndex: Int): ArrayList<Train> {
        var result = ArrayList<Train>()
        try {
            loopNum = 0
            var i = 0
            while (i < sortBefore.size) {
                if (trainList[sortBefore[i]].getPredictionTime(stationIndex) > 0 && !trainList[sortBefore[i]].checkDoubleDay()) {
                    //今からsortAfterに追加する列車の基準駅の時間
                    val baseTime = trainList[sortBefore[i]].getPredictionTime(stationIndex)
                    var j: Int
                    j = sortAfter.size
                    while (j > 0) {
                        if (trainList[sortAfter[j - 1]].getPredictionTime(stationIndex) < baseTime) {
                            break
                        }
                        j--
                    }
                    sortAfter.add(j, sortBefore[i])
                    sortBefore.removeAt(i)
                    i--
                }
                i++
            }
            sorted[stationIndex] = true
            if (direction == Train.Companion.DOWN) {
                sortDown(stationIndex)
            } else {
                sortUp(stationIndex)
            }
            sortAfter.addAll(sortBefore)
            for (i in sortAfter) {
                result.add(trainList[i])
            }
            return result
        } catch (e: Exception) {
            SDlog.log(e)
        }
        result = ArrayList()
        for (i in sortBefore) {
            result.add(trainList[i])
        }
        return result
    }

    /**
     * 路線内を上り方向に探索していきます。
     */
    @Throws(Exception::class)
    fun sortUp(stationIndex: Int) {
        var stationIndex = stationIndex
        loopNum++
        if (loopNum > 50) {
            SDlog.toast("エラーこのダイヤファイルの路線分岐が複雑であるため、列車の並び替え時に無限ループに陥りました。並び替え操作を強制終了します")
            throw Exception("並び替えエラー：" + lineFile.name)
        }
        var skip = true //ソート済みの路線から外れ、別の路線に入る場合skipfragがtrueになる。//ソート済み領域に戻ればskip=false
        while (stationIndex >= 0) {

            //上り方向に探索
            if (sorted[stationIndex]) {
                skip = false
            }
            if (lineFile.getStation(stationIndex)!!.brunchCoreStationIndex >= 0) {
                //この駅が分岐駅設定されている場合
                if (lineFile.getStation(stationIndex)!!.brunchCoreStationIndex > stationIndex) {
                    //下から分岐する場合はソート対象外
                    skip = true
                }
            }
            if (!sorted[stationIndex]) {
                //この駅がまだソートされていないとき
                if (skip) {
                    var subskip = true
                    //この駅がスキップされるとき
                    if (lineFile.getStation(stationIndex)!!.brunchCoreStationIndex >= 0) {
                        //分岐駅設定あり
                        if (sorted[lineFile.getStation(stationIndex)!!.brunchCoreStationIndex]) {
                            //分岐元がソート済み
                            //ソートする
                            subskip = false
                        }
                    }
                    //この駅がどこかの分岐元でその駅がソートされている可能性
                    for (i in 0 until lineFile.stationNum) {
                        if (lineFile.getStation(i)!!.brunchCoreStationIndex == stationIndex && sorted[i]) {
                            subskip = false
                        }
                    }
                    if (subskip) {
                        //スキップする
                        stationIndex--
                        continue
                    }
                }
                //ソートする
                println("sortUp:\t" + stationIndex + "\t(" + lineFile.getStation(stationIndex)!!.name + ")")
                val stations = ArrayList<Int>()
                stations.add(stationIndex)
                if (lineFile.getStation(stationIndex)!!.brunchCoreStationIndex >= 0) {
                    stations.add(lineFile.getStation(stationIndex)!!.brunchCoreStationIndex)
                }
                for (i in 0 until lineFile.stationNum) {
                    if (lineFile.getStation(i)!!.brunchCoreStationIndex == stationIndex) {
                        stations.add(i)
                    }
                }
                if (direction == Train.Companion.DOWN) {
                    addTrainInSort1(stations)
                } else {
                    addTrainInSort2(stations)
                }
                sorted[stationIndex] = true
                skip = false
            }
            if (lineFile.getStation(stationIndex)!!.brunchCoreStationIndex >= 0) {
                //この駅が分岐駅設定されている場合
                if (lineFile.getStation(stationIndex)!!.brunchCoreStationIndex < stationIndex) {
                    //上へ分岐するときは次の駅からソート対象外
                    skip = true
                }
            }
            stationIndex--
        }

        //ループが終わったのに、既にソート済みの駅が最後に残った場合や最後スキップされていた場合
        for (i in 0 until lineFile.stationNum) {
            if (!sorted[i]) {
                sortDown(0)
                return
            }
        }
    }

    /**
     * 路線内を下り方向に探索していきます。
     */
    @Throws(Exception::class)
    fun sortDown(stationIndex: Int) {
        var stationIndex = stationIndex
        loopNum++
        if (loopNum > 50) {
            SDlog.toast("エラーこのダイヤファイルの路線分岐が複雑であるため、列車の並び替え時に無限ループに陥りました。並び替え操作を強制終了します")
            throw Exception("並び替えエラー：" + lineFile.name)
        }
        var skip = true //ソート済みの路線から外れ、別の路線に入る場合skipfragがtrueになる。//ソート済み領域に戻ればskip=false
        while (stationIndex < lineFile.stationNum) {

            //下り方向に探索
            if (sorted[stationIndex]) {
                skip = false
            }
            if (lineFile.getStation(stationIndex)!!.brunchCoreStationIndex >= 0) {
                //この駅が分岐駅設定されている場合
                if (lineFile.getStation(stationIndex)!!.brunchCoreStationIndex < stationIndex) {
                    //上から分岐する場合はソート対象外
                    skip = true
                }
            }
            if (!sorted[stationIndex]) {
                //この駅がまだソートされていないとき
                if (skip) {
                    var subskip = true
                    //この駅がスキップされるとき
                    if (lineFile.getStation(stationIndex)!!.brunchCoreStationIndex >= 0) {
                        //分岐駅設定あり
                        if (sorted[lineFile.getStation(stationIndex)!!.brunchCoreStationIndex]) {
                            //分岐元がソート済み
                            //ソートする
                            subskip = false
                        }
                    }
                    //この駅がどこかの分岐元でその駅がソートされている可能性
                    for (i in 0 until lineFile.stationNum) {
                        if (lineFile.getStation(i)!!.brunchCoreStationIndex == stationIndex && sorted[i]) {
                            subskip = false
                        }
                    }
                    if (subskip) {
                        //スキップする
                        stationIndex++
                        continue
                    }
                }
                //ソートする
                println("sortDown:\t" + stationIndex + "\t(" + lineFile.getStation(stationIndex)!!.name + ")")
                val stations = ArrayList<Int>()
                stations.add(stationIndex)
                if (lineFile.getStation(stationIndex)!!.brunchCoreStationIndex >= 0) {
                    stations.add(lineFile.getStation(stationIndex)!!.brunchCoreStationIndex)
                }
                for (i in 0 until lineFile.stationNum) {
                    if (lineFile.getStation(i)!!.brunchCoreStationIndex == stationIndex) {
                        stations.add(i)
                    }
                }
                if (direction == Train.Companion.DOWN) {
                    addTrainInSort2(stations)
                } else {
                    addTrainInSort1(stations)
                }
                sorted[stationIndex] = true
                skip = false
            }
            if (lineFile.getStation(stationIndex)!!.brunchCoreStationIndex >= 0) {
                //この駅が分岐駅設定されている場合
                if (lineFile.getStation(stationIndex)!!.brunchCoreStationIndex > stationIndex) {
                    //下へ分岐するときは次の駅からソート対象外
                    skip = true
                }
            }
            stationIndex++
        }
        //ループが終わったのに、既にソート済みの駅が最後に残った場合や最後スキップされていた場合
        for (i in 0 until lineFile.stationNum) {
            if (!sorted[i]) {
                sortUp(lineFile.stationNum - 1)
                break
            }
        }
    }

    /**
     * 列車をsortAfterに時刻前方から挿入する
     * station[0]に停車する列車がソート対象
     * station[1以上]は同一駅
     */
    private fun addTrainInSort1(station: ArrayList<Int>) {
        for (i in sortBefore.size downTo 1) {
            val baseTime = trainList[sortBefore[i - 1]].getTime(station[0], Train.Companion.ARRIVE, true)
            if (baseTime < 0 || trainList[sortBefore[i - 1]].checkDoubleDay()) {
                continue
            }
            var j = 0
            var frag = false
            j = 0
            while (j < sortAfter.size) {
                var sortTime = -1
                for (s in station) {
                    sortTime = Math.max(sortTime, trainList[sortAfter[j]].getPredictionTime(s, Train.Companion.DEPART))
                }
                if (sortTime < 0) {
                    j++
                    continue
                }
                frag = true
                if (sortTime >= baseTime) {
                    break
                }
                j++
            }
            if (frag) {
                sortAfter.add(j, sortBefore[i - 1])
                sortBefore.removeAt(i - 1)
            }
        }
    }

    private fun addTrainInSort2(station: ArrayList<Int>) {
        var i = 0
        while (i < sortBefore.size) {
            val baseTime = trainList[sortBefore[i]].getDepTime(station[0])
            if (baseTime < 0 || trainList[sortBefore[i]].checkDoubleDay()) {
                i++
                continue
            }
            var j = 0
            var frag = false
            j = sortAfter.size - 1
            while (j >= 0) {
                var sortTime = -1
                for (s in station) {
                    sortTime = Math.max(sortTime, trainList[sortAfter[j]].getPredictionTime(s, Train.Companion.ARRIVE))
                }
                if (sortTime < 0) {
                    j--
                    continue
                }
                frag = true
                if (sortTime <= baseTime) {
                    break
                }
                j--
            }
            if (frag) {
                sortAfter.add(j + 1, sortBefore[i])
                sortBefore.removeAt(i)
                i--
            }
            i++
        }
    }

    internal inner class TrainNumberSorter(var train: Train) : Comparable<TrainNumberSorter> {
        var name = ""
        var number = 0
        fun isNumber(c: Char): Boolean {
            return c >= '0' && c <= '9'
        }

        override fun compareTo(o: TrainNumberSorter): Int {
            return if (name == o.name) {
                number - o.number
            } else name.compareTo(o.name)
        }

        init {
            val value = train.number
            for (i in 0 until value!!.length) {
                if (!isNumber(value.toCharArray()[i])) {
                    if (i != 0) {
                        number = value.substring(0, i).toInt()
                    }
                    name = value.substring(i)
                    return
                }
            }
            if (value.length != 0) {
                number = value.toInt()
            }
        }
    }

    internal inner class TrainTypeSorter(var train: Train) : Comparable<TrainTypeSorter> {
        override fun compareTo(o: TrainTypeSorter): Int {
            return train.type - o.train.type
        }

    }

    internal inner class TrainNameSorter(var train: Train) : Comparable<TrainNameSorter> {
        override fun compareTo(o: TrainNameSorter): Int {
            return if (train.name == o.train.name) {
                train.count.compareTo(o.train.count)
            } else train.name.compareTo(o.train.name)
        }

    }

    internal inner class TrainRemarkSorter(var train: Train) : Comparable<TrainRemarkSorter> {
        override fun compareTo(o: TrainRemarkSorter): Int {
            return train.remark.compareTo(o.train.remark)
        }

    }

    /**
     * ソートする時刻表を入力します。
     */
    init {
        sortBefore = ArrayList()
        sortAfter = ArrayList()
        for (i in trainList.indices) {
            sortBefore.add(i)
        }
        sorted = BooleanArray(lineFile.stationNum)
        for (i in sorted.indices) {
            sorted[i] = false
        }
    }
}